horovod/_keras/__init__.py

```diff
@@ -13,22 +13,76 @@
-            self._get_gradients_used = False
-            super(self.__class__, self).__init__(**kwargs)
+def _allreduce_gradients(gradients, name, sparse_as_dense, device_dense, device_sparse,
+                         compression):
+    if hvd.size() > 1:
+        averaged_gradients = []
+        with tf.name_scope(name + "_Allreduce"):
+            for grad in gradients:
+                if grad is not None:
+                    if sparse_as_dense and \
+                            isinstance(grad, tf.IndexedSlices):
+                        grad = tf.convert_to_tensor(grad)
+                    avg_grad = hvd.allreduce(grad,
+                                             device_dense=device_dense,
+                                             device_sparse=device_sparse,
+                                             compression=compression)
+                    averaged_gradients.append(avg_grad)
+                else:
+                    averaged_gradients.append(None)
+            return averaged_gradients
+    else:
+        return gradients
+
+
+def _create_distributed_optimizer_class(tf_version_above_2, keras, name, device_dense,
+                                        device_sparse, compression, sparse_as_dense):
+    def _create_distributed_optimizer_base_class(base_class, method_name):
+        class _DistributedOptimizerBase(base_class):
+            def __init__(self, **kwargs):
+                self._name = name or "Distributed%s" % self.__class__.__base__.__name__
+                self._device_dense = device_dense
+                self._device_sparse = device_sparse
+                self._compression = compression
+                self._sparse_as_dense = sparse_as_dense
+                self._get_gradient_or_get_unscaled_gradients_called = False
+                super(self.__class__, self).__init__(**kwargs)
+
+            def apply_gradients(self, *args, **kwargs):
+                if not self._get_gradient_or_get_unscaled_gradients_called:
+                    raise Exception('`apply_gradients()` was called without a call to '
+                                    '`{method_name}()`.'.format(method_name=method_name))
+                return super(self.__class__, self).apply_gradients(*args, **kwargs)
+
+        return _DistributedOptimizerBase
+
+    if tf_version_above_2:
+        def get_unscaled_gradients(self, *args, **kwargs):
+            """
+            Unscales the gradients by the loss scale.
 
+            See LossScaleOptimizer.get_unscaled_gradients() for more info.
+
+            In DistributedOptimizer, get_unscaled_gradients() is overriden to
+            allreduce the gradients before returning them.
+            """
+            self._get_gradient_or_get_unscaled_gradients_called = True
+            gradients = super(self.__class__, self).get_unscaled_gradients(*args, **kwargs)
+            return _allreduce_gradients(gradients, self._name, self._sparse_as_dense,
```

Comment:
I believe we would need to do the allreduce before getting the unscaled gradients to be consistent with past behavior.  @DEKHTIARJonathan can confirm.